src/IECoreMaya/LiveScene.cpp

```diff
@@ -512,6 +517,185 @@ void LiveScene::writeTags( const NameList &tags )
+	typedef CurvesMergeOp MergeOpType;
+	static CurvesPrimitiveParameter * primParameter( MergeOpType* mop )
+	{
+		return mop->curvesParameter();
+	}
+};
+
+template<>
+struct PrimMergerTraits<MFn::kMesh>
+{
+	typedef FromMayaMeshConverter ConverterType;
+	typedef MeshMergeOp MergeOpType;
+	static MeshPrimitiveParameter * primParameter( MergeOpType* mop )
+	{
+		return mop->meshParameter();
+	}
+};
+
+template<int MFnType>
+class PrimMerger
+{
+		typedef typename PrimMergerTraits<MFnType>::ConverterType ConverterType;
+		typedef typename PrimMergerTraits<MFnType>::MergeOpType MergeOpType;
+		typedef typename MergeOpType::PrimitiveType PrimitiveType;
+
+	public:
+
+		static void createMergeOp( MDagPath &childDag, IECore::ModifyOpPtr &op )
+		{
+			typename ConverterType::Ptr converter = runTimeCast<ConverterType>( ConverterType::create( childDag ) );
+			if( ! converter )
+			{
+				throw Exception( ( boost::format( "Creating merge op failed! " ) % childDag.fullPathName().asChar() ).str() );
+			}
+			typename PrimitiveType::Ptr cprim = runTimeCast<PrimitiveType>( converter->convert() );
+			op = new MergeOpType;
+			op->copyParameter()->setTypedValue( false );
+			op->inputParameter()->setValue( cprim );
+		}
+
+		static void mergePrim( MDagPath &childDag, IECore::ModifyOpPtr &op )
+		{
+			typename ConverterType::Ptr converter = runTimeCast<ConverterType>( ConverterType::create( childDag ) );
+			if( ! converter )
+			{
+				throw Exception( ( boost::format( "Merging primitive failed! " ) % childDag.fullPathName().asChar() ).str() );
+			}
+			typename PrimitiveType::Ptr prim = runTimeCast<PrimitiveType>( converter->convert() );
+			MergeOpType *mop = runTimeCast<MergeOpType>( op.get() );
+			PrimMergerTraits<MFnType>::primParameter( mop )->setValue( const_cast<PrimitiveType*>( prim.get() ) );
+			op->operate();
+		}
+
+};
+
+bool hasMergeableObjects( const MDagPath &p )
+{
+	// When there are multiple child shapes that can be merged, readMergedObject() returns an object that has all the shapes merged in it.
+	// This is because multiple Maya shapes can be converted to one IECore primitive eg. nurbs curves -> IECore::CurvesPrimitive.
+	// We want to have multiple shape nodes in Maya, and want it to be one primitive if viewed through IECoreMaya::LiveScene.
+	unsigned int childCount = p.childCount();
+
+	// At least two shapes need to exist to merge.
+	if( childCount < 2 )
+	{
+		return MFn::kInvalid;
+	}
+
+	MFn::Type foundType = MFn::kInvalid;
+	for ( unsigned int c = 0; c < childCount; c++ )
+	{
+		MObject childObject = p.child( c );
+		MFn::Type type = childObject.apiType();
+
+		if( type == MFn::kMesh || type == MFn::kNurbsCurve )
+		{
+			if( MFnDagNode( childObject ).isIntermediateObject() )
+			{
+				continue;
+			}
+
+			if( foundType == MFn::kInvalid )
+			{
+				foundType = type;
+			}
+			else if( foundType == type )
+			{
+				// Found a mergeable shape with the same type as before.
+				return true;
```

Comment:
we shouldn't be returning true early. we should be returning false early if `foundType != type`